src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -542,15 +577,84 @@ internal static async Task<Document> CleanupSyntaxAsync(Document document, CodeC
 
     internal static async Task<Document> CleanupSemanticsAsync(Document document, CodeCleanupOptions options, CancellationToken cancellationToken)
     {
-        var document1 = await ImportAdder.AddImportsFromSymbolAnnotationAsync(document, Simplifier.AddImportsAnnotation, options.AddImportOptions, cancellationToken).ConfigureAwait(false);
-        var document2 = await Simplifier.ReduceAsync(document1, Simplifier.Annotation, options.SimplifierOptions, cancellationToken).ConfigureAwait(false);
-        var document3 = await CaseCorrector.CaseCorrectAsync(document2, CaseCorrector.Annotation, cancellationToken).ConfigureAwait(false);
+        var newSolution = await CleanupSemanticsAsync(
+            document.Project.Solution, [(document.Id, options)], CodeAnalysisProgress.None, cancellationToken).ConfigureAwait(false);
+        return newSolution.GetRequiredDocument(document.Id);
+    }
+
+    internal static async Task<Solution> CleanupSemanticsAsync(
+        Solution solution,
+        ImmutableArray<(DocumentId documentId, CodeCleanupOptions options)> documentIdsAndOptions,
+        IProgress<CodeAnalysisProgress> progress,
+        CancellationToken cancellationToken)
+    {
+        // We do this in 4 serialized passes.  This allows us to process all documents in parallel, while only forking
+        // the solution 4 times *total* (instead of 4 times *per* document).
+        progress.AddItems(documentIdsAndOptions.Length * 4);
```

Comment:
this is the meat of hte change.  4 defined passes on the solution.  not 4 passes per document.